### PR TITLE
edk2_basic:Add edk2 basic test

### PR DESCRIPTION
**Pull request by zhenyzha**
Check whether the status of edk2
The boot process is normal.

ID: 2574

### CODE REVIEW

**Comment by zhenyzha** [comment]
@XueqiangWei @PaulYuuu Could you pls help review when you are free. thanks.
After testing, the patch can effectively reproduce the Assertion and Disconnect issues.

```
(598/600) edk2_basic.arm64-pci: PASS (316.10 s)
(599/600) edk2_basic.iommu_test.smmuv3.arm64-pci: PASS (418.24 s)
(600/600) edk2_basic.iommu_test.viommu.arm64-pci: PASS (323.35 s)
RESULTS    : PASS 590 | ERROR 0 | FAIL 10 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```

File: qemu/tests/edk2_basic.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    vm = env.get_vm(params["main_vm"])
+
+    for i in range(params.get_numeric("reboot_count", 1)):
+        vm.create()
+        error_context.context("Check serial log result", test.log.info)
+        try:
+            vm.serial_console.read_until_output_matches([check_messgae],
+                                                        timeout=timeout)
+            error_context.context("Check edk2 output information", test.log.info)
+            output = vm.serial_console.get_output()
```
Comment (PaulYuuu) [inline qemu/tests/edk2_basic.py:31]: `read_until_output_matches` will return output, and the final line matches the `[check_message]`, better than getting new data from get_output.

File: qemu/tests/cfg/edk2_basic.cfg [hunk 2]
```diff
@@ -0,0 +1,20 @@
+- edk2_basic:
+    virt_test_type = qemu libvirt
```
Comment (PaulYuuu) [inline qemu/tests/cfg/edk2_basic.cfg:2]:
```suggestion
    virt_test_type = qemu
```

File: qemu/tests/edk2_basic.py (hunk 1 above)
Comment (zhenyzha) [reply qemu/tests/edk2_basic.py:31]: The output here is to capture all logs to match the number of lines. The keyword matching has been used above with `read_until_output_matches`

File: qemu/tests/cfg/edk2_basic.cfg (hunk 2 above)
Comment (zhenyzha) [reply qemu/tests/cfg/edk2_basic.cfg]: Corrected

File: qemu/tests/edk2_basic.py (hunk 1 above)
Comment (PaulYuuu) [reply qemu/tests/edk2_basic.py:31]: I don't understand re-getting the output here. the output of `read_until_output_matches` returns lines before `check_message`, serial log prints very quickly,  `vm.serial_console.get_output()` may get more lines after `check_message`, which is not accurate to use to judge.

File: qemu/tests/cfg/edk2_basic.cfg [hunk 3]
```diff
@@ -0,0 +1,20 @@
+- edk2_basic:
+    virt_test_type = qemu
+    only Linux
+    type = edk2_basic
```
Comment (PaulYuuu) [inline qemu/tests/cfg/edk2_basic.cfg:4]: For edk2, only available for aarch64 and x86 q35.

File: qemu/tests/cfg/edk2_basic.cfg [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    start_vm = no
+    login_timeout = 240
+    reboot_count = 5
+    check_messgae = "enter to boot the selected OS"
+    skip_image_processing = yes
+    variants:
+        - @default:
+        - iommu_test:
```
Comment (PaulYuuu) [inline qemu/tests/cfg/edk2_basic.cfg:12]: x86 has intel-iommu.

Comment (zhenyzha) [reply qemu/tests/cfg/edk2_basic.cfg:14]: Corrected

File: qemu/tests/cfg/edk2_basic.cfg (hunk 3 above)
Comment (zhenyzha) [reply qemu/tests/cfg/edk2_basic.cfg:5]: Corrected

File: qemu/tests/edk2_basic.py (hunk 1 above)
Comment (zhenyzha) [reply qemu/tests/edk2_basic.py]: Corrected

**Comment by zhenyzha** [comment]
@XueqiangWei @PaulYuuu Could you pls help review again when you are free? thanks.

Will fail due to product bug cases:
`edk2_basic.arm64-pci: FAIL: Warning edk2 line count exceeds 40`

You can manually change `line_numbers = 400` to verify the script
```
 (1/3) edk2_basic.arm64-pci: PASS (79.15 s)
 (2/3) edk2_basic.iommu_test.smmuv3.arm64-pci: PASS (79.63 s)
 (3/3) edk2_basic.iommu_test.viommu.arm64-pci: PASS (79.46 s)
```

**Comment by XueqiangWei** [comment]
(1/3) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.edk2_basic.q35: STARTED
 (1/3) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.edk2_basic.q35: PASS (106.36 s)
 (2/3) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.edk2_basic.iommu_test.intel_iommu.q35: STARTED
 (2/3) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.edk2_basic.iommu_test.intel_iommu.q35: PASS (94.91 s)
 (3/3) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.edk2_basic.iommu_test.viommu.q35: STARTED
 (3/3) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.5.0.x86_64.io-github-autotest-qemu.edk2_basic.iommu_test.viommu.q35: PASS (93.73 s)

File: qemu/tests/cfg/edk2_basic.cfg [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+                    machine_type_extra_params = "kernel-irqchip=split"
+                    virtio_dev_ats = on
+                    virtio_dev_aer = on
+                    intel_iommu = yes
+                    enable_guest_iommu = yes
+                - smmuv3:
+                    only aarch64
+                    machine_type_extra_params += ",iommu=smmuv3"
+                - viommu:
+                    only aarch64 q35
```
Comment (PaulYuuu) [inline qemu/tests/cfg/edk2_basic.cfg:30]: EDK2 means OVMF in x86 and only supports q35, so this limitation is unnecessary if you have the correct definition on the top.

File: qemu/tests/cfg/edk2_basic.cfg [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+                    no Host_RHEL.m9.u0
+                    required_qemu= [7.0.0,)
+                    virtio_iommu = yes
+                    virtio_iommu_direct_plug = yes
```
Comment (PaulYuuu) [inline qemu/tests/cfg/edk2_basic.cfg:34]: This option is dropped.

Comment (zhenyzha) [reply qemu/tests/cfg/edk2_basic.cfg]: Thanks review, updated.

File: qemu/tests/cfg/edk2_basic.cfg (hunk 5 above)
Comment (zhenyzha) [reply qemu/tests/cfg/edk2_basic.cfg]: Thanks review, updated

**Comment by zhenyzha** [comment]
@XueqiangWei @PaulYuuu Could you pls help review again or ack when you are free? thanks.

File: qemu/tests/cfg/edk2_basic.cfg [hunk 7]
```diff
@@ -0,0 +1,32 @@
+- edk2_basic:
+    virt_test_type = qemu
+    only Linux
+    only x86_64, aarch64
```
Comment (PaulYuuu) [inline qemu/tests/cfg/edk2_basic.cfg:4]:
```suggestion
    only x86_64, aarch64
    x86_64:
        only ovmf
```

File: qemu/tests/cfg/edk2_basic.cfg [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    skip_image_processing = yes
+    line_numbers = 40
+    variants:
+        - @default:
+        - iommu_test:
+            virtio_dev_iommu_platform = on
+            virtio_dev_filter = '^(?:(?:virtio-)|(?:vhost-))(?!(?:balloon)|(?:user)|(?:iommu))'
+            variants:
+                - intel_iommu:
+                    only q35
```
Comment (PaulYuuu) [inline qemu/tests/cfg/edk2_basic.cfg:19]:
```suggestion
                    only x86_64
```

File: qemu/tests/cfg/edk2_basic.cfg (hunk 7 above)
Comment (zhenyzha) [reply qemu/tests/cfg/edk2_basic.cfg:4]: Oh sorry I didn't understand your suggestion correctly before, many thanks for the example, I have corrected it.

File: qemu/tests/cfg/edk2_basic.cfg (hunk 8 above)
Comment (zhenyzha) [reply qemu/tests/cfg/edk2_basic.cfg]: Already corrected

**Comment by zhenyzha** [comment]
@XueqiangWei @PaulYuuu Could you pls help review again or ack when you are free? thanks.

**Review by PaulYuuu — APPROVED** [review]

**Comment by XueqiangWei** [comment]
Tested on x86 platform, all passed. 
(1/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.x86_64.io-github-autotest-qemu.edk2_basic.q35: STARTED
 (1/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.x86_64.io-github-autotest-qemu.edk2_basic.q35: PASS (80.55 s)
 (2/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.x86_64.io-github-autotest-qemu.edk2_basic.iommu_test.intel_iommu.q35: STARTED
 (2/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.x86_64.io-github-autotest-qemu.edk2_basic.iommu_test.intel_iommu.q35: PASS (80.92 s)
 (3/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.x86_64.io-github-autotest-qemu.edk2_basic.iommu_test.viommu.q35: STARTED
 (3/3) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.x86_64.io-github-autotest-qemu.edk2_basic.iommu_test.viommu.q35: PASS (80.24 s)

**Comment by XueqiangWei** [comment]
@zhenyzha Suggest updating the commit message, adding more detailed description for all checkpoints. Thanks.

**Comment by zhenyzha** [comment]
@XueqiangWei Already updated, could you please help review again or ack when you are free? thanks.

**Review by XueqiangWei — APPROVED** [review]
LGTM,
Acked-by: xuwei@redhat.com

**Comment by XueqiangWei** [comment]
@zhenyzha commit signature verification has been enabled, please re-push it with -S. Thanks.

**Comment by zhenyzha** [comment]
> @zhenyzha commit signature verification has been enabled, please re-push it with -S. Thanks.

@XueqiangWei  Hello Xueqiang, Thanks for the reminder, it has been added.

**Review by XueqiangWei — APPROVED** [review]

**Pull request merged**